File: 06-Arrays-Hashing/cpp/minimum_window_substring.cpp
/*
 * Problem: Minimum Window Substring
 * Source:  LeetCode 76 - https://leetcode.com/problems/minimum-window-substring/
 * 
 * Approach: Expand-shrink sliding window with frequency map. Expand right to
 *           include all chars of t, then shrink left to minimize window.
 * Time Complexity:  O(|s| + |t|)
 * Space Complexity: O(|t|)
 * 
 * Tags: arrays, hashing, sliding-window, two-pointers
 */

#include <bits/stdc++.h>
using namespace std;

string minWindow(string s, string t) {
    unordered_map<char, int> need, window;
    for (char c : t) need[c]++;
    
    int left = 0, right = 0;
    int valid = 0; // number of characters satisfied
    int start = 0, minLen = INT_MAX;
    
    while (right < (int)s.size()) {
        char c = s[right];
        right++;
        
        if (need.count(c)) {
            window[c]++;
            if (window[c] == need[c]) valid++;
        }
        
        // Try to shrink window
        while (valid == (int)need.size()) {
            if (right - left < minLen) {
                start = left;
                minLen = right - left;
            }
            
            char d = s[left];
            left++;
            
            if (need.count(d)) {
                if (window[d] == need[d]) valid--;
                window[d]--;
            }
        }
    }
    
    return minLen == INT_MAX ? "" : s.substr(start, minLen);
}

int main() {
    cout << "Min window: \"" << minWindow("ADOBECODEBANC", "ABC") << "\"" << endl;
    // Output: "BANC"
    
    cout << "Min window: \"" << minWindow("a", "a") << "\"" << endl;
    // Output: "a"
    
    return 0;
}

File: 06-Arrays-Hashing/java/MinimumWindowSubstring.java
/*
 * Problem: Minimum Window Substring
 * Source:  LeetCode 76 - https://leetcode.com/problems/minimum-window-substring/
 * 
 * Approach: Expand-shrink sliding window with frequency map.
 * Time Complexity:  O(|s| + |t|)
 * Space Complexity: O(|t|)
 * 
 * Tags: arrays, hashing, sliding-window, two-pointers
 */

import java.util.*;

public class MinimumWindowSubstring {
    
    public String minWindow(String s, String t) {
        Map<Character, Integer> need = new HashMap<>(), window = new HashMap<>();
        for (char c : t.toCharArray()) need.merge(c, 1, Integer::sum);
        
        int left = 0, right = 0, valid = 0;
        int start = 0, minLen = Integer.MAX_VALUE;
        
        while (right < s.length()) {
            char c = s.charAt(right);
            right++;
            
            if (need.containsKey(c)) {
                window.merge(c, 1, Integer::sum);
                if (window.get(c).equals(need.get(c))) valid++;
            }
            
            while (valid == need.size()) {
                if (right - left < minLen) {
                    start = left;
                    minLen = right - left;
                }
                
                char d = s.charAt(left);
                left++;
                
                if (need.containsKey(d)) {
                    if (window.get(d).equals(need.get(d))) valid--;
                    window.merge(d, -1, Integer::sum);
                }
            }
        }
        
        return minLen == Integer.MAX_VALUE ? "" : s.substring(start, start + minLen);
    }
    
    public static void main(String[] args) {
        MinimumWindowSubstring sol = new MinimumWindowSubstring();
        System.out.println("Min window: \"" + sol.minWindow("ADOBECODEBANC", "ABC") + "\""); // BANC
        System.out.println("Min window: \"" + sol.minWindow("a", "a") + "\""); // a
    }
}
